### ISSUE TITLE
Wrong position for images used/shared by several produtcs

### ISSUE THREAD

**Issue by PierreLeMaguer** [magento/magento2#17325]
### Preconditions

1. Magento CE 2.1.14

### Steps to reproduce

1. 2 simple products
2. these 2 products have to share the same image. In the media tables, it means : 
in catalog_product_entity_media_gallery_value
```
| value_id | store_id | entity_id | label | position | disabled | record_id |
|    24045 |        0 |      7113 | NULL  |        2 |        0 |     92279 |
|    24045 |        0 |      2283 | NULL  |        6 |        0 |     92328 |
```
and in catalog_product_entity_media_gallery
```
| value_id | attribute_id | value                      | media_type | disabled |
|    24045 |           90 | /n/a/nanatsunotaizai_3.png | image      |        0 |
```
and in catalog_product_entity_media_gallery_value_to_entity
```
| value_id | entity_id |
|    24045 |      2283 |
|    24045 |      7113 |
```
### Expected result

1. Images are displayed with the right position for each products.

### Actual result

1. Images shared are not displayed at the right position.

### Additional Comments
Well, Magento 2 allow to share images for several products thanks to its media table definition. I did not find a BO way to share one image for several products. I do it only with a custom import process.
I find a solution by editing this method : 
vendor/magento/module-catalog/Model/ResourceModel/Product/Gallery.php :: createBatchBaseSelect by adding 1 condition AND in the two joined left tables. The select becomes : 
```
$select = $this->getConnection()->select()->from(
            [$mainTableAlias => $this->getMainTable()],
            [
                'value_id',
                'file' => 'value',
                'media_type'
            ]
        )->joinInner(
            ['entity' => $this->getTable(self::GALLERY_VALUE_TO_ENTITY_TABLE)],
            $mainTableAlias . '.value_id = entity.value_id',
            [$linkField]
        )->joinLeft(
            ['value' => $this->getTable(self::GALLERY_VALUE_TABLE)],
            implode(
                ' AND ',
                [
                    $mainTableAlias . '.value_id = value.value_id',
                    $this->getConnection()->quoteInto('value.store_id = ?', (int)$storeId),
                    'entity.entity_id = value.entity_id' // Edit
                ]
            ),
            ['label', 'position', 'disabled']
        )->joinLeft(
            ['default_value' => $this->getTable(self::GALLERY_VALUE_TABLE)],
            implode(
                ' AND ',
                [
                    $mainTableAlias . '.value_id = default_value.value_id',
                    $this->getConnection()->quoteInto('default_value.store_id = ?', Store::DEFAULT_STORE_ID),
                    'entity.entity_id = default_value.entity_id' // Edit
                ]
            ),
            ['label_default' => 'label', 'position_default' => 'position', 'disabled_default' => 'disabled']
        )->where(
            $mainTableAlias . '.attribute_id = ?',
            $attributeId
        )->where(
            $mainTableAlias . '.disabled = 0'
        )->order(
            $positionCheckSql . ' ' . \Magento\Framework\DB\Select::SQL_ASC
        );
```
When I find time, I will make a PR.
Thanks.

**Comment by magento-engcom-team** [comment]
Hi @PierreLeMaguer. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [x] Summary of the issue
- [x] Information on your environment
- [x] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me {$VERSION} instance` 

where `{$VERSION}` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (2.2-develop +). 
For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation.

@PierreLeMaguer do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [x] no

**Comment by ghost** [comment]
@PierreLeMaguer, thank you for your report.
We were not able to reproduce this issue by following the steps you provided. If you'd like to update it, please reopen the issue.

**Comment by PierreLeMaguer** [comment]
@magento-engcom-team give me 2.2-develop instance

**Comment by magento-engcom-team** [comment]
Hi @PierreLeMaguer. Thank you for your request. I'm working on Magento 2.2-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @PierreLeMaguer, here is your [Magento instance](https://i-17325-2-2-develop.engcom.dev.magento.com).
Admin access: [https://i-17325-2-2-develop.engcom.dev.magento.com/admin](https://i-17325-2-2-develop.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by PierreLeMaguer** [comment]
@magento-engcom-team Can I have sql access ?

**Comment by PierreLeMaguer** [comment]
The issue is only present on Magento 2.1.X and was corrected on 2.2.X and 2.3.X . If you are facing this issue, my solution works or you can see the magento solution on the same class in the 2.2-dev branch.
